grant-project/zcash-grant-system
# Require refund address or confirm donation

Closes #321.

### What This Does

* Adds a `no_refund` column to contributions that mark them as not to be refunded
    * Should an admin choose, they can refund anyway using the same process
* Requires users either have a refund address or confirm that it'll be a donation before contributing
* Show donations in financials page
    * Donations include transactions that should be refunded, but are anonymous or no_refund
    * Also reworked financials page to be more responsive
* Adds API-side check to ensure refund address isn't cleared out, only changed

### Steps to Test

* Run `flask db upgrade`
* Create a new proposal & new user (or existing user w/o refund address)
* Make a contribution to the proposal as the user, confirm you got a prompt to set a refund address
* Choose to make it a donation and make the contribution. Confirm it has `no_refund == True`
* Make another contribution of the same amount, but this time enter a refund address. Confirm it has `no_refund == False`
* Attempt to make a third contribution, confirm it no longer asks you about refund address.
* Complete the first two contributions (either through admin or sending)
* Cancel and refund the proposal, confirm that only the contribution that is `no_refund == False` prompts admins for refund
* Go to your settings page, confirm you can only change the refund address, not unset it
    * Not sure if this is still the desired behavior with the other changes, but this is how it was
* Check the admin financials page to make sure the numbers add up
* Check the admin contributions list page with the Donation filter on to make sure it shows the right contributions

## Screenshots

### Refund address prompt

<img width="547" alt="Screen Shot 2019-03-13 at 6 14 01 PM" src="https://user-images.githubusercontent.com/649992/54318186-d66a4380-45bb-11e9-9d15-cb49e5068d3e.png">


### Admin financials

<img width="1437" alt="Screen Shot 2019-03-13 at 6 05 05 PM" src="https://user-images.githubusercontent.com/649992/54318051-68be1780-45bb-11e9-9e52-ec71fd88e7ab.png">
